src/riak_dt_hll.erl

```diff
@@ -0,0 +1,376 @@
+-export([to_binary/2]).
+-export([stats/1, stat/2]).
+-export([parent_clock/2]).
+-export([to_version/2]).
+
+%% Additional APIs for hyperloglog types
+-export([new/1, reduce/2, precision/1, check_precision_and_reduce/2]).
+
+-export_type([hllset/0, card/0, precision/0]).
+
+-define(HYPER_DEFAULT_BACKEND, hyper_binary).
+-define(STATS, [bytes, precision]).
+
+-define(DT_HLLSET_TAG, 86).
+-define(TAG, ?DT_HLLSET_TAG).
+-define(V1_VERS, 1).
+
+%% Error macros pulled from riak_dt_tags.hrl
+-define(UNSUPPORTED_VERSION(Vers), {error, unsupported_version, Vers}).
+-define(INVALID_BINARY, {error, invalid_binary}).
+-define(UNSUPPORTED_VERSION, {error, unsupported_version,
+                              Vers :: pos_integer()}).
+-define(OPS, [add, add_all]).
+
+-record(hyper, {p :: precision(),
+                registers :: {module(), registers()}}).
+
+-opaque hllset() :: #hyper{}.
+
+-type hllset_op() :: {add|add_all, term()|list(term())} |
+                     {update, [hllset_op()]}.
+
+-type precision() :: hyper:precision().
+-type registers() :: hyper:registers().
+-type card() :: number().
+-type actor() :: riak_dt:actor().
+-type stat() :: bytes | precision.
+
+-spec new() -> hllset().
+new() ->
+    new(?HYPER_DEFAULT_PRECISION).
+
+-spec new(precision()|riak_kv_bucket:props()) -> hllset().
+new(Precision) when is_integer(Precision), Precision > 3
+                    andalso Precision < 17 ->
+    hyper:new(Precision, ?HYPER_DEFAULT_BACKEND);
+new(BProps) when is_list(BProps) ->
+    Precision = proplists:get_value(hll_precision, BProps,
+                                    ?HYPER_DEFAULT_PRECISION),
+    new(Precision);
+new(_) ->
+    new(?HYPER_DEFAULT_PRECISION).
+
+-spec reduce(hllset(), precision()) -> hllset().
+reduce(HllSet, NewPrecision) ->
+    hyper:reduce_precision(NewPrecision, HllSet).
+
+-spec precision(hllset()) -> precision().
+precision(HllSet) ->
+    hyper:precision(HllSet).
+
+-spec value(hllset()) -> card().
+value(HllSet) ->
+    round(hyper:card(HllSet)).
+value(precise, HllSet) ->
+    value(HllSet).
+
+-spec update(hllset_op(), actor(), hllset()) -> {ok, hllset()} |
+                                               {error, term()}.
+update({add, Elem}, _Actor, HllSet) ->
+    {ok, hyper:insert(Elem, HllSet)};
+update({add_all, Elems}, _Actor, HllSet) ->
+    {ok, hyper:insert_many(Elems, HllSet)};
+update({update, Ops}, Actor, HllSet) ->
+    apply_ops(Ops, Actor, HllSet).
+
+-spec update(hllset_op(), actor(), hllset(), riak_dt:context())
+            -> {ok, hllset()}.
+update(Op, Actor, HllSet, _Ctx) ->
+    update(Op, Actor, HllSet).
+
+-spec parent_clock(riak_dt_vclock:vclock(), hllset()) -> hllset().
+parent_clock(_Clock, HllSet) ->
+    HllSet.
+
+-spec merge(hllset(), hllset()) -> hllset().
+merge(HllSet1, HllSet2) ->
+    hyper:union(HllSet1, HllSet2).
+
+-spec equal(hllset(), hllset()) -> boolean().
+equal(HllSet1, HllSet2) ->
+    hyper:to_json(HllSet1) == hyper:to_json(HllSet2).
```

Comment:
is the best way to check equality? Is it expensive? Wondering where riak might call this (I don't know off hand.)